src/service/framerate-impl.js

```diff
@@ -0,0 +1,163 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {assert} from '../asserts';
+import {getService} from '../service';
+import {timer} from '../timer';
+import {viewerFor} from '../viewer';
+
+const collectTime = 5000;
+
+export class Framerate {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window}  */
+    this.win = win;
+
+    /**
+     * Return value of requestAnimationFrame for use with
+     * cancelRequestAnimationFrame.
+     * @private {number}
+     */
+    this.requestedFrame_ = null;
+
+    /**
+     * Whether we've scheduled a frame already. Maintained
+     * to avoid having to request and cancel all the time.
+     * @private {boolean}
+     */
+    this.frameScheduled_ = false;
+
+    /** @private {number}  */
+    this.lastFrameTime_ = 0;
+
+    /** @private {number}  */
+    this.collectUntilTime_ = 0;
+
+    /** @private {number}  */
+    this.collectStartTime_ = 0;
+
+    /** @private {number}  */
+    this.frameCount_ = 0;
+
+    /**
+     * Whether we loaded an ad on this page.
+     * @private {boolean}
+     */
+    this.loadedAd_ = false;
+
+    const viewer = viewerFor(this.win);
+
+    /**
+     * We do not make measurements when the window is hidden, because
+     * animation frames not not fire in that case.
+     * @private {boolean}
+     */
+    this.isActive_ = viewer.isVisible();
+
+    viewer.onVisibilityChanged(() => {
+      this.isActive_ = viewer.isVisible();
+      if (this.isActive_) {
+        this.reset_();
+        this.collect();
+        console.info('visible')
+      } else {
+        this.win.cancelAnimationFrame(this.requestedFrame_);
+        this.frameScheduled_ = false;
+        this.reset_();
+        console.info('hidden')
+      }
+    });
+
+    this.collect();
+  }
+
+  /**
+   * Call this when something interesting is about to happen on screen.
+   * This class will then measure the framerate for the next few seconds.
+   * @param {!Element=} opt_element Element for which the current
+   *     collection is requested.
+   */
+  collect(opt_element) {
+    if (!this.isActive_ || !this.win.requestAnimationFrame) {
+      return;
+    }
+    const now = timer.now();
+    if (this.lastFrameTime_ == 0) {
+      this.collectStartTime_ = now;
+    }
+    if (opt_element && opt_element.tagName == 'AMP-AD') {
+      this.loadedAd_ = true;
+    }
+    this.collectUntilTime_ = now + collectTime;
+    this.collect_(now);
+  }
+
+  reset_() {
+    this.frameCount_ = 0;
```

Comment:
The idea is to taint everything when the ad was ever loaded. Since it might effect all subsequent frames.